### PR TITLE
Add com.chenyifaer.FaFaRunner

### PR DESCRIPTION
**Pull request by kjxbyz**
### Please confirm your submission meets all the criteria

- [X] Please describe the application briefly.

A game developed using flame and bonfire.

- [X] The domain used for the application ID is [controlled by the application developers][appid-domain] and the [application id guidelines][appid] are followed.
- [X] I have read the and followed all the [Submission and licence requirements][reqs].
- [X] I have [built][build] and tested the submission locally.
- [X] I am an author/developer/upstream contributor of the project. If not, I contacted upstream developers about this submission. **Link:**

[appid-domain]: https://docs.flathub.org/docs/for-app-authors/requirements/#control-over-domain-or-repository
[appid]: https://docs.flathub.org/docs/for-app-authors/requirements#application-id
[reqs]: https://docs.flathub.org/docs/for-app-authors/requirements
[build]: https://docs.flathub.org/docs/for-app-authors/submission/#before-submission

### CODE REVIEW

**Comment by bbhtt** [comment]
bot, build com.chenyifaer.FaFaRunner

**Comment by bbhtt** [comment]
screenshot links are 404 https://raw.githubusercontent.com/fafarunner/fafarunner/v1.0.0+355/screenshots/desktop.png

You can start the build once that is fixed

File: com.chenyifaer.FaFaRunner.yml [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+        dest-filename: fafarunner.deb
+        sha256: 84dbd660fe4c3e76091470314cea23b843a2fc8551fb2656275e972f7decfba5
+        x-checker-data:
+          type: json
+          url: https://api.github.com/repos/fafarunner/fafarunner/releases/latest
+          version-query: .tag_name | sub("v";"")
+          url-query: .assets[] | select(.name=="fafarunner_" + $version + "_linux_aarch64.deb") | .browser_download_url
+
+      - type: file
+        url: https://raw.githubusercontent.com/fafarunner/fafarunner/v1.0.0+355/linux/shared/com.chenyifaer.FaFaRunner.desktop
```
Comment (hfiguiere) [inline com.chenyifaer.FaFaRunner.yml:57]: this is not in the .deb? it should be.

File: com.chenyifaer.FaFaRunner.yml [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+        dest-filename: com.chenyifaer.FaFaRunner.desktop
+        sha256: eefde8dc8a139ae8102a43bebaea7f2334076cbf685ae1b5243f0b3496cff252
+        x-checker-data:
+          type: json
+          url: https://api.github.com/repos/fafarunner/fafarunner/releases/latest
+          version-query: .tag_name
+          url-query: "\"https://raw.githubusercontent.com/fafarunner/fafarunner/\" + $version + \"/linux/shared/com.chenyifaer.FaFaRunner.desktop\""
+
+      - type: file
+        url: https://raw.githubusercontent.com/fafarunner/fafarunner/v1.0.0+355/linux/shared/com.chenyifaer.FaFaRunner.metainfo.xml
```
Comment (hfiguiere) [inline com.chenyifaer.FaFaRunner.yml:67]: this is not in the .deb? it should be.

File: com.chenyifaer.FaFaRunner.yml (hunk 1 above)
Comment (kjxbyz) [reply com.chenyifaer.FaFaRunner.yml:57]: This file is universal and is also used by other Linux format packages.

File: com.chenyifaer.FaFaRunner.yml (hunk 2 above)
Comment (kjxbyz) [reply com.chenyifaer.FaFaRunner.yml:67]: This file is universal and is also used by other Linux format packages.

File: com.chenyifaer.FaFaRunner.yml (hunk 1 above)
Comment (hfiguiere) [reply com.chenyifaer.FaFaRunner.yml:57]: then it should be in the .deb

File: com.chenyifaer.FaFaRunner.yml (hunk 2 above)
Comment (hfiguiere) [reply com.chenyifaer.FaFaRunner.yml:67]: then it should be in the .deb

**Comment by kjxbyz** [comment]
The screenshots are missing, it will be fixed in the next version.

File: com.chenyifaer.FaFaRunner.yml (hunk 1 above)
Comment (hfiguiere) [reply com.chenyifaer.FaFaRunner.yml]: this is still unresolved.

Comment (kjxbyz) [reply com.chenyifaer.FaFaRunner.yml]: You need to check Files changed

File: com.chenyifaer.FaFaRunner.yml [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+  - --share=network
+
+modules:
+  - name: fafarunner
+    buildsystem: simple
+    build-commands:
+      - bsdtar -Oxf fafarunner.deb data.tar.zst | bsdtar -xf -
+      - install -Dm755 usr/share/fafarunner/fafarunner /app/bin/fafarunner
+      - cp -r usr/share/fafarunner/data usr/share/fafarunner/lib /app/bin/
+      - sed -i -e '/^Version=/d' -e 's|Icon=fafarunner|Icon=com.chenyifaer.FaFaRunner|' usr/share/applications/fafarunner.desktop
```
Comment (hfiguiere) [inline com.chenyifaer.FaFaRunner.yml:27]: don't use sed. Use the command `desktop-file-edit`.

File: com.chenyifaer.FaFaRunner.yml [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+      - install -Dm644 usr/share/applications/fafarunner.desktop /app/share/applications/com.chenyifaer.FaFaRunner.desktop
+      - install -Dm644 usr/share/icons/hicolor/128x128/apps/fafarunner.png /app/share/icons/hicolor/128x128/apps/com.chenyifaer.FaFaRunner.png
+      - install -Dm644 usr/share/icons/hicolor/256x256/apps/fafarunner.png /app/share/icons/hicolor/256x256/apps/com.chenyifaer.FaFaRunner.png
+      - sed -i -e 's|<id>fafarunner</id>|<id>com.chenyifaer.FaFaRunner</id>|' -e 's|fafarunner.desktop|com.chenyifaer.FaFaRunner.desktop|' usr/share/metainfo/fafarunner.metainfo.xml
```
Comment (hfiguiere) [inline com.chenyifaer.FaFaRunner.yml:31]: don't use sed. It doesn't have a failure mode.

Note that if you keep the desktop file with its original name and use `rename-desktop-file` in the manifest, then it's done for you. Same with the appstream file using `rename-appdata-file` and the icons using `rename-icon`.

Comment (kjxbyz) [reply com.chenyifaer.FaFaRunner.yml:31]: Sorry, I don't understand

Comment (kjxbyz) [reply com.chenyifaer.FaFaRunner.yml:31]: The manifest file is universal. Will it have any impact on other Linux distribution packages?

Please provide an example.

Comment (bbhtt) [reply com.chenyifaer.FaFaRunner.yml:31]: > The manifest file is universal. Will it have any impact on other Linux distribution packages?

No `fafarunner` is an incorrect ID. It needs to be rDNS. It'd be better if you directly fix these issues in the .deb.

Add the old id `fafarunner` in the `provides` tag of metainfo and it should be fine:

```
<provides>
  <id>fafarunner</id>
</provides>

Comment (kjxbyz) [reply com.chenyifaer.FaFaRunner.yml:31]: Maybe this?

```
<id>com.chenyifaer.FaFaRunner</id>
<provides>
  <id>fafarunner</id>
</provides>
```

Comment (bbhtt) [reply com.chenyifaer.FaFaRunner.yml:31]: Yes

File: com.chenyifaer.FaFaRunner.yml (hunk 3 above)
Comment (kjxbyz) [reply com.chenyifaer.FaFaRunner.yml:27]: Which package contains the command `desktop-file-edit`?

Comment (bbhtt) [reply com.chenyifaer.FaFaRunner.yml:27]: desktop-file-utils on your distro, but it's in the runtime by default. You don't need to add it, it works out of the box

File: com.chenyifaer.FaFaRunner.yml (hunk 4 above)
Comment (kjxbyz) [reply com.chenyifaer.FaFaRunner.yml:31]: Can I understand it this way? I added these tags, and what happens at runtime is:
File rename
1. `fafarunner.png` -> `com.chenyifaer.FaFaRunner.png`
2. `fafarunner.desktop` -> `com.chenyifaer.FaFaRunner.desktop`
3. `fafarunner.metainfo.xml` -> `com.chenyifaer.FaFaRunner.metainfo.xml`

File content changes
1. The value of `Icon` in the `fafarunner.desktop` file is updated from `fafarunner` to `com.chenyifaer.FaFaRunner`
2. The value of `id` in the `fafarunner.metainfo.xml` file is updated from `fafarunner.desktop` to `com.chenyifaer.FaFaRunner.desktop`
3. The value of `launchable` in the `fafarunner.metainfo.xml` file is updated from `fafarunner.desktop` to `com.chenyifaer.FaFaRunner.desktop`

Comment (kjxbyz) [reply com.chenyifaer.FaFaRunner.yml:31]: In addition, the `.desktop` file is automatically generated by the packaging tool, and it contains some wrong fields. I have to use sed to remove these wrong fields.

Comment (hfiguiere) [reply com.chenyifaer.FaFaRunner.yml:31]: That's what `desktop-file-edit` is for. Not `sed`.

Comment (kjxbyz) [reply com.chenyifaer.FaFaRunner.yml:31]: Can `desktop-file-edit` automatically remove invalid fields in `.desktop` files?

Comment (kjxbyz) [reply com.chenyifaer.FaFaRunner.yml:31]: I think it might be more suitable for me to move the `.desktop` and `.metainfo.xml` files to this repository.

Comment (hfiguiere) [reply com.chenyifaer.FaFaRunner.yml:31]: No.

Comment (kjxbyz) [reply com.chenyifaer.FaFaRunner.yml:31]: I just need to change
```
<id>fafarunner</id>
```

to
```
<id>com.chenyifaer.FaFaRunner</id>
<provides>
  <id>fafarunner</id>
</provides>
```

and then no other changes are needed?

File: com.chenyifaer.FaFaRunner.yml (hunk 3 above)
Comment (kjxbyz) [reply com.chenyifaer.FaFaRunner.yml:27]: Maybe this?

```
desktop-file-edit --remove-key='Version' --set-icon='com.chenyifaer.FaFaRunner' "usr/share/applications/fafarunner.desktop"
```

**Comment by kjxbyz** [comment]
bot, build com.chenyifaer.FaFaRunner

**Comment by kjxbyz** [comment]
bot, build com.chenyifaer.FaFaRunner

**Pull request closed without merging**